### DISCUSSION TITLE
Can openEMS be used for metamaterials? And if yes, do you have any tips?

### DISCUSSION THREAD

**Discussion by plebcoding** [thliebig/openEMS-Project#208] [Q&A Usage]
Hello everyone 

I am an undergraduate physics student and I found out about this open-source software while I was looking for simulation tools that I could use for my research. I am very new to EM analysis software and I was hoping that I could use this software in order to run simulations for the metamaterial design that I plan to do. Specifically, I would want to record the properties of a metamaterial as it interacts with EM waves in the NIR to Visible wavelength; properties such as absorption, transmission, reflection, dispersion, etc. in both TM and TE modes. 

I am new to coding in general and most of my experience with coding is with MATLAB for my programming and physics classes, however I do have a bit of experience with Python. So if it was possible I would really appreciate it if I could use openEMS hand-in-hand with MATLAB as I am more comfortable using it than Python. Another issue is that I really am new to coding and I dont have a lot of experience with it, which makes me anxious if openEMS is feasible for someone who isn't fluent with coding and is still learning it. So here are my questions regarding openEMS, 

1. Can I use this to simulate metamaterials? 
2. Can I extract the properties that I want by using openEMS?
3. Is openEMS good to use hand-in-hand with MATLAB? 
4. Can someone who isn't that good with coding learn to use openEMS?

I am really excited to explore using openEMS and hopefully it could be a tool that I can use for my undergraduate research. I am having fun exploring the Discussion board here and how there are a lot of resources available, hopefully some people would be able to help me with the questions I have regarding openEMS. 

I guess that would be it for now and of course thank you so much in advance to those who would reply to my post!

**Comment by georgy-moshkin** [comment]
1. Yes, you can simulate metamaterials. E.g., I use it to simulate mushroom-type metamaterials. Very simple periodic structure. Basically, it is patch with ground via in center (google for papers on mushroom type metamaterials). Patch can be rectangular, circular, hexagonal, etc. Btw, I have an online course on this and other antennas which uses openEMS.
2. openEMS comes with many useful *.m Octave scripts which extract all required data from openEMS FDTD output. Time domain EM-fields, function to extract frequency-domain data (S-parameters, far-field, etc..). The same stuff as any commercial EM simulator
3. You can read raw openEMS FDTD solver output files using MATLAB functions and do some processing based on *.m scripts provided with openEMS. But it's already working out of the box with Octave and everything is free.
4. Of course. Even 30 years ago, many researchers used QBasic, Fortran and other tools for scientific calculations. In some cases, code for calculating patch antenna included small FDTD/MOM solver or something like that. Now, it is much easier, with google's help.

![image](https://github.com/thliebig/openEMS-Project/assets/48346176/c7f2ab50-9e97-423f-8b10-80459c9ed4b1)

**Reply by plebcoding** [reply]
Hey thanks a lot for the answer, it seems like it would be worth it to explore openEMS and using it for my research. I already have MATLAB so im going to definitely start exploring this simulation tool. Based on the pictures that you have attached it seems similar to the outputs that I need in order to simulate metamaterials. 

BTW it looks interesting! This is the first time I have heard of mushroom metamaterials so I had no prior knowledge about it before I saw your reply, I might also explore that category of metamaterials and maybe there is something there that I could use! 

Also, you mentioned that you had some online lectures on implementing openEMS, would it be possible to view these lectures? I think it would be very useful for me as someone who will be trying out openEMS for the first time, and I would be very thankful if I could watch/read the online lectures that you have regarding the use of openEMS. Nonetheless, thank you so much for your answer to my question! I'm glad I could use this software for metamaterials and I'm getting excited just thinking about exploring this new tool for simulations lol.

**Reply by georgy-moshkin** [reply]
Some chapters on antenna design will be available on my GitHub this summer.

For a quick start: 
1) Extract openEMS so it is placed in C:\openEMS\
2) Install Octave
3) Run Octave, Edit → Set Path → click Add Folder → Single folder → select C:\openEMS\matlab, click "Reload"
4) Editor tab, From file menu click "Open...", and open examples from C:\openEMS\matlab\examples and C:\openEMS\matrlab\tutorials
5) Drag command window tab to the left, so you can always see if *.m file asks you "Are you sure to delete files, type YES".
6) Click "play" button ("Save file and Run"). Note that for most examples you need to type "yes" and press enter in command window. On a first run Octave asks "Change directory?" - answer yes.

**Comment by biergaizi** [comment]
1. Can I use this to simulate metamaterials?

Yes. openEMS already contains a [simulation example](https://docs.openems.de/python/openEMS/Tutorials/CRLH_Extraction.html) of a planar antenna based on Composite Right/Left Handed (CRLH) transmission lines, which is a type of metamaterial. 

> I would want to record the properties of a metamaterial as it interacts with EM waves in the NIR to Visible wavelength

Unfortunately openEMS is not designed or tested in the optical regime, only in the microwave spectrum for radio engineering. I think nobody uses openEMS above 60 GHz and whether simulations would even run correctly at optical frequencies is an open question. It also only contains Drude and Lorentz material models, alternative models for nanotechnology or optics is non-existent. Thus, it's probably *not the correct choice* if IR/visible light simulation is a hard requirement. However, if studying material properties at microwave frequencies is good enough, it may be useful for you.

2. Can I extract the properties that I want by using openEMS?

Yes. Once openEMS produces the S-parameters for the antenna, it's possible to find its equivalent circuits or other parameters by curve fitting. However, openEMS only calculates the raw physics (frequency response, time-domain waveforms, E&M fields, radiation patterns). It's up to you to do the curve-fitting math via standard numerical methods, openEMS can't do this automatically for you.

3. Is openEMS good to use hand-in-hand with MATLAB?

Yes. You can call openEMS directly from MATLAB.

4. Can someone who isn't that good with coding learn to use openEMS?

Programming knowledge is required to use openEMS. You need your own code for creating the 3D models, for starting simulations, and for post-processing. But exactly how much expertise is required, depends on the complexity of your simulations. A big problem is that openEMS lacks ready-made 3D models for common objects, and each has to be modeled from scratch using basic geometry. If you're only simulating planar objects or 3D objects with simple shapes like a dipole, the programming effort is not too high. On the other hand, if your object is 3D with irregular shapes or many details, the 3D modeling part itself is already a challenge.
